soc/arm/st_stm32/stm32l4/power.c

```diff
@@ -0,0 +1,168 @@
+		LL_PWR_DisableSRAM3Retention();
+		/* enter STOP2 mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP2);
+		LL_LPM_EnableDeepSleep();
+		k_cpu_idle();
+		break;
+ #endif
+#endif /* CONFIG_SYS_POWER_SLEEP_STATES */
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP_STATES
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_DEEP_SLEEP_1
+	case SYS_POWER_STATE_DEEP_SLEEP_1:
+		/* this corresponds to the STANDBY mode: */
+#ifdef CONFIG_DEBUG
+		/* Enable the Debug Module during Standy mode */
+		LL_DBGMCU_EnableDBGStandbyMode();
+#endif
+		/* enable Power clock */
+		LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+		/* clear WUFx bits in the power status register 1 */
+		LL_PWR_ClearFlag_WU();
+		/* the content of SRAM2 will be lost */
+		LL_PWR_DisableSRAM2Retention();
+		/* set all GPIO to analog mode and disable GPIO clocks */
+		/* TODO */
+		/* program the wakeUp source RTC or wdg or wakeUp pin */
+		/* TODO */
+		/* enter STANDBY mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STANDBY);
+		LL_LPM_EnableDeepSleep();
+		k_cpu_idle();
+		break;
+ #endif
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_DEEP_SLEEP_2
+	case SYS_POWER_STATE_DEEP_SLEEP_2:
+		/* this corresponds to the SHUTDOWN mode */
+		/* enable Power clock */
+		LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+		/* clear WUFx bits in the power status register 1 */
+		LL_PWR_ClearFlag_WU();
+		/* set all GPIO to analog mode and disable GPIO clocks */
+		/* TODO */
+		/* program the wakeUp source RTC or wdg or wakeUp pin */
+		/* TODO */
+		/* enter SHUTDOWN mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_SHUTDOWN);
+		LL_LPM_EnableDeepSleep();
+		k_cpu_idle();
+		break;
+ #endif
+#endif /* CONFIG_SYS_POWER_DEEP_SLEEP_STATES */
+	default:
+		LOG_DBG("Unsupported power state %u", state);
+		break;
+	}
+}
+
+/* Handle SOC specific activity after Low Power Mode Exit */
+void _sys_pm_power_state_exit_post_ops(enum power_states state)
+{
+	switch (state) {
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_1
+	case SYS_POWER_STATE_SLEEP_1:
+		LL_LPM_DisableSleepOnExit();
+		break;
+ #endif
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_2
+	case SYS_POWER_STATE_SLEEP_2:
+		LL_LPM_DisableSleepOnExit();
+		break;
+ #endif
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_3
+	case SYS_POWER_STATE_SLEEP_3:
+		LL_LPM_DisableSleepOnExit();
+		break;
+ #endif
+#endif /* CONFIG_SYS_POWER_SLEEP_STATES */
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP_STATES
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_DEEP_SLEEP_1
+	case SYS_POWER_STATE_DEEP_SLEEP_1:
+		LL_PWR_DisableLowPowerRunMode();
+		break;
+ #endif
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_DEEP_SLEEP_2
+	case SYS_POWER_STATE_DEEP_SLEEP_2:
+		LL_PWR_DisableLowPowerRunMode();
+		break;
+ #endif
+#endif /* CONFIG_SYS_POWER_DEEP_SLEEP_STATES */
+	default:
+		LOG_DBG("Unsupported power state %u", state);
+		break;
+	}
+
+	/*
+	 * System is now in active mode.
+	 * Reenable interrupts which were disabled
+	 * when OS started idling code.
+	 */
+	irq_unlock(0);
+}
```

Comment:
Do you still work on this pull request?
I take your code for my project. I restart the clock on wakeup like this
```
struct device* dev = device_get_binding(STM32_CLOCK_CONTROL_NAME);
SystemCoreClock = 160000000;
dev->config->init(dev);
```